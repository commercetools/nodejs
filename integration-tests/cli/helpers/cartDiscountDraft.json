{
  "name": {
    "en": "integration-test-cart-discount"
  },
  "value": {
    "type": "relative",
    "permyriad": 1000
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "target": {
    "type": "lineItems",
    "predicate": "1 = 1"
  },
  "sortOrder": "0.1",
  "requiresDiscountCode": true
}
