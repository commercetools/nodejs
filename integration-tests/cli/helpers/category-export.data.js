export default [
  {
    key: 'fooKey',
    name: {
      en: 'categoryA',
    },
    slug: {
      en: 'category-a',
    },
    description: {
      en: 'descA',
    },
    externalId: 'a',
    assets: [],
  },
  {
    key: 'barKey',
    name: {
      en: 'categoryB',
    },
    slug: {
      en: 'category-b',
    },
    description: {
      en: 'descB',
    },
    externalId: 'b',
    assets: [],
  },
  {
    key: 'fooBarKey',
    name: {
      en: 'categoryC',
    },
    slug: {
      en: 'category-c',
    },
    description: {
      en: 'descC',
    },
    externalId: 'c',
    assets: [],
  },
]
