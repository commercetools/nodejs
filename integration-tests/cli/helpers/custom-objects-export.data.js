export default [
  {
    container: 'Ludus',
    key: 'copperKey',
    value: {
      paymentMethod: 'Cash',
      paymentID: '1',
    },
  },
  {
    container: 'Frobozz',
    key: 'jadeKey',
    value: {
      paymentMethod: 'cc',
      paymentID: '2',
    },
  },
  {
    container: 'Syrinx',
    key: 'crystalKey',
    value: {
      paymentMethod: 'new',
      paymentID: '3',
    },
  },
]
