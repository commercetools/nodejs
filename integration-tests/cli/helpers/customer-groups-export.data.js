export default [
  {
    groupName: 'Ludus',
    key: 'copperKey',
  },
  {
    groupName: 'Frobozz',
  },
  {
    groupName: 'Syrinx',
    key: 'crystalKey',
  },
]
