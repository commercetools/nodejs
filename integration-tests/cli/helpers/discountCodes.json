[{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "IT8MKKZX"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITlN85n0"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITDHzsVX"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITlb6Z1u"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITfWgUVd"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITpQgE1P"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITL2pvbZ"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITlrnkj5"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITDJO3LT"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITuIcDI2"
}]
