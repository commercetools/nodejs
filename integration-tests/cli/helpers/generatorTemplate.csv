name.en,name.de,description.en,description.de,cartPredicate,isActive,maxApplications,maxApplicationsPerCustomer
Sammy,Valerian,greatest promo,super angebot,some cart predicate,true,9,4
