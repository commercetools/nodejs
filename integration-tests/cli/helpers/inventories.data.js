export const inventories = [{
  sku: '12345',
  quantityOnStock: 20,
  custom: {
    type: {
      key: 'custom-type',
    },
    fields: {
      description: 'integration tests!! arrgggh',
    },
  },
}]

export const customFields = [{
  key: 'custom-type',
  name: {
    en: 'customized fields',
  },
  description: {
    en: 'custom description',
  },
  resourceTypeIds: ['inventory-entry'],
  fieldDefinitions: [{
    name: 'description',
    type: {
      name: 'String',
    },
    required: false,
    label: {
      en: 'owner',
    },
    inputHint: 'SingleLine',
  }],
}]
