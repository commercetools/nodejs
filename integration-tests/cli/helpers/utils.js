import { createAuthMiddlewareForClientCredentialsFlow } from '@commercetools/sdk-middleware-auth'
import { createClient } from '@commercetools/sdk-client'
import { createRequestBuilder } from '@commercetools/api-request-builder'
import { createHttpMiddleware } from '@commercetools/sdk-middleware-http'
import fetch from 'node-fetch'

export function clearData(apiConfig, entityName) {
  const client = createClient({
    middlewares: [
      createAuthMiddlewareForClientCredentialsFlow({ ...apiConfig, fetch }),
      createHttpMiddleware({ host: apiConfig.apiUrl, fetch }),
    ],
  })

  const service = createRequestBuilder({
    projectKey: apiConfig.projectKey,
  })[entityName]

  const request = {
    uri: service.build(),
    method: 'GET',
  }
  return client.process(request, payload => {
    // Built-in states cannot be deleted
    const results =
      entityName === 'states'
        ? payload.body.results.filter(state => state.builtIn === false)
        : payload.body.results
    return Promise.all(
      results.map(result =>
        client.execute({
          uri: service
            .byId(result.id)
            .withVersion(result.version)
            .build(),
          method: 'DELETE',
        })
      )
    )
  })
}

export function createData(apiConfig, entityName, data, id) {
  const client = createClient({
    middlewares: [
      createAuthMiddlewareForClientCredentialsFlow({ ...apiConfig, fetch }),
      createHttpMiddleware({ host: apiConfig.apiUrl, fetch }),
    ],
  })
  const requestOption = { projectKey: apiConfig.projectKey }
  const service = createRequestBuilder(requestOption)[entityName]
  return Promise.all(
    data.map(_data => {
      if (id) service.byId(id)
      const request = {
        uri: service.build(),
        method: 'POST',
        body: _data,
      }
      return client.execute(request)
    })
  )
}

export function getId(apiConfig, entityName) {
  const client = createClient({
    middlewares: [
      createAuthMiddlewareForClientCredentialsFlow({ ...apiConfig, fetch }),
      createHttpMiddleware({ host: apiConfig.apiUrl, fetch }),
    ],
  })

  const service = createRequestBuilder({
    projectKey: apiConfig.projectKey,
  })[entityName]

  const request = {
    uri: service.build(),
    method: 'GET',
  }

  return client
    .execute(request)
    .then(result => Promise.resolve(result.body.results[0].id))
}
