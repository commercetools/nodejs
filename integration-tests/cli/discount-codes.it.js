import fs from 'mz/fs'
import tmp from 'tmp'
import path from 'path'
import isuuid from 'isuuid'
import csv from 'csvtojson'
import { exec } from 'mz/child_process'
import fetch from 'node-fetch'
import DiscountCodeImport from '@commercetools/discount-code-importer'
import { getCredentials } from '@commercetools/get-credentials'
import { createData, clearData } from './helpers/utils'

let projectKey
if (process.env.CI === 'true') projectKey = 'discount-codes-integration-test'
else projectKey = process.env.npm_config_projectkey

describe('DiscountCode tests', () => {
  let apiConfig
  let cartDiscount

  const logger = {
    error: () => {},
    warn: () => {},
    info: () => {},
    verbose: () => {},
  }

  beforeAll(async () => {
    const credentials = await getCredentials(projectKey)
    // Get test credentials

    apiConfig = {
      host: 'https://auth.europe-west1.gcp.commercetools.com',
      apiUrl: 'https://api.europe-west1.gcp.commercetools.com',
      projectKey,
      credentials,
    }

    // Clear all discount codes
    await clearData(apiConfig, 'discountCodes')
    // Clear all cart discounts
    await clearData(apiConfig, 'cartDiscounts')

    // Create cart-discount
    const cartDiscountDraft = await fs.readFileSync(
      path.join(__dirname, './helpers/cartDiscountDraft.json'),
      'utf8'
    )
    // Wrap in an array because the util function expects an array
    const data = await createData(apiConfig, 'cartDiscounts', [
      cartDiscountDraft,
    ])

    cartDiscount = data[0].body

  }, 15000)

  afterAll(async () => {
    // Delete discount codes
    await clearData(apiConfig, 'discountCodes')
    // Delete cart discounts
    await clearData(apiConfig, 'cartDiscounts')
  })

  describe('Discount code generator', () => {
    const binPath = './integration-tests/node_modules/.bin/discount-code-gen'

    // Delete the generated log file
    afterAll(() => {
      fs.unlinkSync('discountCodeGenerator.log', 'utf8')
    })

    it('should generate required codes according to template', async () => {
      const expected = {
        name: { en: 'Sammy', de: 'Valerian' },
        description: { en: 'greatest promo', de: 'super angebot' },
        cartPredicate: 'some cart predicate',
        isActive: true,
        maxApplications: 9,
        maxApplicationsPerCustomer: 4,
      }
      const filePath = path.join(__dirname, './helpers/generatorTemplate.csv')

      const [stdout, stderr] = await exec(
        `${binPath} -q 10 -p IT -l 8 -i ${filePath}`
      )
      expect(stderr).toBeFalsy()
      const generatedCodes = JSON.parse(stdout)
      expect(generatedCodes).toHaveLength(10)
      generatedCodes.forEach((codeObj) => {
        expect(codeObj).toMatchObject(expected)
        expect(codeObj.code).toMatch(/^IT/)
        expect(codeObj.code).toHaveLength(8)
      })
    }, 15000)
  })

  describe('Discount code Importer', () => {
    let codeImport
    let preparedDiscountCodes
    beforeEach(() => {
      codeImport = new DiscountCodeImport({ apiConfig }, logger)

      // Get the sample discount codes and add cartDiscounts
      preparedDiscountCodes = JSON.parse(
        fs.readFileSync(
          path.join(__dirname, './helpers/discountCodes.json'),
          'utf8'
        )
      ).map((codeObj) => ({
        ...codeObj,
        cartDiscounts: [
          {
            typeId: 'cart-discount',
            id: cartDiscount.id,
          },
        ],
      }))
    }, 20000)

    // Delete Discount codes
    afterAll(() => clearData(apiConfig, 'discountCodes', fetch))

    it('should create discount codes on CTP', async () => {
      await codeImport.run(preparedDiscountCodes)
      const summary = codeImport.summaryReport()
      expect(summary).toMatchSnapshot()
    }, 15000)

    it('should update discount codes on the CTP', async () => {
      // First, import the codes that need to be updated
      const oldCodesToUpdate = preparedDiscountCodes.map((codeObj) => {
        const uniqueCode = codeObj.code
        return { ...codeObj, code: `${uniqueCode}foo` }
      })
      await codeImport.run(oldCodesToUpdate)

      // Call a new `codeImport` instance so we reset the summary
      codeImport = new DiscountCodeImport({ apiConfig }, logger)

      const newCodesToUpdate = oldCodesToUpdate.map((codeObj) => ({
        ...codeObj,
        maxApplications: 20,
      }))

      await codeImport.run(newCodesToUpdate)
      const summary = codeImport.summaryReport()
      expect(summary).toMatchSnapshot()
    })

    it('should stop import on first error by default', async () => {
      // Set batchSize to 1 so it executes serially
      codeImport = new DiscountCodeImport({ apiConfig, batchSize: 1 }, logger)
      // Make codes unique
      const discountCodesSample = preparedDiscountCodes.map((codeObj) => {
        const uniqueCode = codeObj.code
        return { ...codeObj, code: `${uniqueCode}bar` }
      })

      // Make code invalid
      discountCodesSample[1].code = ''
      discountCodesSample[2].code = ''

      try {
        await codeImport.run(discountCodesSample)
      } catch (e) {
        expect(e.summary.created).toBe(1)
        expect(e.summary.errors).toHaveLength(1)
        expect(e.summary.errors[0]).toMatch(/'code' should not be empty/)
      }
    })

    it('should continueOnProblems if `continueOnProblems`', async () => {
      codeImport = new DiscountCodeImport(
        {
          apiConfig,
          batchSize: 5,
          continueOnProblems: true,
        },
        logger
      )
      // Make codes unique
      const discountCodesSample = preparedDiscountCodes.map((codeObj) => {
        const uniqueCode = codeObj.code
        return { ...codeObj, code: `${uniqueCode}foobar` }
      })
      // Make code invalid
      discountCodesSample[1].code = ''
      discountCodesSample[7].cartDiscounts = 'INVALID-CART-DISCOUNT'

      await codeImport.run(discountCodesSample)
      const summary = codeImport.summaryReport()
      expect(summary).toMatchSnapshot()
      const errors = summary.detailedSummary.errors
      expect(errors).toMatchSnapshot()
    })
  })

  describe('Discount Code Exporter', () => {
    let preparedDiscountCodes
    const bin = './integration-tests/node_modules/.bin/discount-code-exporter'

    beforeAll(async () => {
      await clearData(apiConfig, 'discountCodes')

      // Get the sample discount codes and add cartDiscounts
      preparedDiscountCodes = JSON.parse(
        fs.readFileSync(
          path.join(__dirname, './helpers/discountCodes.json'),
          'utf8'
        )
      ).map((codeObj) => ({
        ...codeObj,
        cartDiscounts: [
          {
            typeId: 'cart-discount',
            id: cartDiscount.id,
          },
        ],
      }))
      await createData(apiConfig, 'discountCodes', preparedDiscountCodes).catch(
        process.stderr.write
      )
    }, 15000)

    it('should write json output to file by default', async () => {
      const jsonFilePath = tmp.fileSync().name
      const expected = {
        version: 1,
        name: {
          en: 'Sammy',
          de: 'Valerian',
        },
        description: {
          en: 'greatest promo',
          de: 'super angebot',
        },
        cartDiscounts: [
          {
            typeId: 'cart-discount',
            id: cartDiscount.id,
          },
        ],
        cartPredicate: 'lineItemTotal(1 = 1) >  "10.00 USD"',
        isActive: true,
        maxApplications: 10,
        maxApplicationsPerCustomer: 2,
      }

      const [stdout, stderr] = await exec(
        `${bin} -o ${jsonFilePath} -p ${projectKey}`
      )
      expect(stderr).toBeFalsy()
      expect(stdout).toMatchSnapshot()

      const data = await fs.readFile(jsonFilePath, { encoding: 'utf8' })

      const actual = JSON.parse(data)
      actual.forEach((codeObj) => {
        expect(codeObj).toMatchObject(expected)
        expect(isuuid(codeObj.id)).toBe(true)
      })
    })

    it('should write csv output to file when passed the option', async () => {
      const csvFilePath = tmp.fileSync().name
      const expected = {
        name: {
          en: 'Sammy',
        },
        description: {
          en: 'greatest promo',
        },
        cartDiscounts: cartDiscount.id,
        cartPredicate: 'lineItemTotal(1 = 1) >  "10.00 USD"',
        isActive: 'true',
        maxApplications: '10',
        maxApplicationsPerCustomer: '2',
      }
      const [stdout, stderr] = await exec(
        `${bin} -o ${csvFilePath} -p ${projectKey} -f csv`
      )
      expect(stderr).toBeFalsy()
      expect(stdout).toMatch(/Export operation completed successfully/)

      const data = await fs.readFile(csvFilePath, { encoding: 'utf8' })

      await new Promise((resolve) => {
        csv()
          .fromString(data)
          .subscribe((jsonObj) => {
            expect(jsonObj).toMatchObject(expected)
            resolve()
          })
      })
    })
  })
})
