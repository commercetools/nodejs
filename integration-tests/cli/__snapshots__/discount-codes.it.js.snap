// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiscountCode tests Discount Code Exporter should write json output to file by default 1`] = `
"info Starting Export
info Export operation completed successfully
"
`;

exports[`DiscountCode tests Discount code Importer should continueOnProblems if \`continueOnProblems\` 1`] = `
Object {
  "detailedSummary": Object {
    "createErrorCount": 2,
    "created": 8,
    "errors": Array [
      "'code' should not be empty.",
      "Request body does not contain valid JSON.",
    ],
    "unchanged": 0,
    "updateErrorCount": 0,
    "updated": 0,
  },
  "reportMessage": "Summary: there were 8 successfully imported discount codes (8 were newly created, 0 were updated and 0 were unchanged). 2 errors occured (2 create errors and 0 update errors.)",
}
`;

exports[`DiscountCode tests Discount code Importer should continueOnProblems if \`continueOnProblems\` 2`] = `
Array [
  "'code' should not be empty.",
  "Request body does not contain valid JSON.",
]
`;

exports[`DiscountCode tests Discount code Importer should create discount codes on CTP 1`] = `
Object {
  "detailedSummary": Object {
    "createErrorCount": 0,
    "created": 10,
    "errors": Array [],
    "unchanged": 0,
    "updateErrorCount": 0,
    "updated": 0,
  },
  "reportMessage": "Summary: there were 10 successfully imported discount codes (10 were newly created, 0 were updated and 0 were unchanged).",
}
`;

exports[`DiscountCode tests Discount code Importer should update discount codes on the CTP 1`] = `
Object {
  "detailedSummary": Object {
    "createErrorCount": 0,
    "created": 0,
    "errors": Array [],
    "unchanged": 0,
    "updateErrorCount": 0,
    "updated": 10,
  },
  "reportMessage": "Summary: there were 10 successfully imported discount codes (0 were newly created, 10 were updated and 0 were unchanged).",
}
`;
